File: Source/NSObject.m
/* gnu/gnustep/base/NSObject.java native code
   Copyright (C) 2000 Free Software Foundation, Inc.
   
   Written by:  Nicola Pero <nicola@brainstorm.co.uk>
   Date: June 2000
   
   This file is part of JIGS, the GNUstep Java Interface Library.

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Library General Public
   License as published by the Free Software Foundation; either
   version 2 of the License, or (at your option) any later version.
   
   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Library General Public License for more details.
   
   You should have received a copy of the GNU Library General Public
   License along with this library; if not, write to the Free
   Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111 USA.
   */ 

#include <jni.h>
#include <Foundation/Foundation.h>
#include "gnu/gnustep/base/NSObject.h"
#include "JIGS.h"

JNIEXPORT jlong JNICALL 
Java_gnu_gnustep_base_NSObject_NSObject_1alloc (JNIEnv *env, jobject this)
{
  Class objectClass;
  id objc;

  JIGS_ENTER;

  objectClass = JIGSClassOfThis (env, this);  
  objc = [objectClass alloc];
  _JIGSMapperAddJavaProxy (env, objc, this);

  JIGS_EXIT_WITH_FAIL_VALUE (0);

  return JIGS_ID_TO_JLONG (objc);
}

JNIEXPORT jlong JNICALL 
Java_gnu_gnustep_base_NSObject_NSObject_1new (JNIEnv *env, jobject this)
{
  Class objectClass;
  id objc;

  JIGS_ENTER;

  objectClass = JIGSClassOfThis (env, this);  
  objc = [objectClass new];
  _JIGSMapperAddJavaProxy (env, objc, this);

  JIGS_EXIT_WITH_FAIL_VALUE (0);

  return JIGS_ID_TO_JLONG (objc);
}

JNIEXPORT jboolean JNICALL 
Java_gnu_gnustep_base_NSObject_equals (JNIEnv *env, jobject this, 
				       jobject other)
{ 
  id we, they;
  jboolean return_value;

  JIGS_ENTER;

  we = JIGSIdFromThis (env, this);
  they =  JIGSIdFromJobject (env, other);
  return_value = (jboolean)[we isEqual: they];

  JIGS_EXIT_WITH_FAIL_VALUE (JNI_FALSE);  

  return return_value;
}

/*
 * The following is special, so we take advantage of a little trick by
 * directly passing the IDpointer as argument and sparing the lookup.  */
JNIEXPORT void JNICALL 
Java_gnu_gnustep_base_NSObject_finalize_1native (JNIEnv *env, jobject this, 
						 jlong IDpointer)
{
  id objc;
  JIGS_ENTER;

  objc = JIGS_JLONG_TO_ID (IDpointer);

  _JIGSMapperRemoveJavaProxy (env, objc);

  RELEASE (objc);

  JIGS_EXIT;
}


JNIEXPORT jint JNICALL 
Java_gnu_gnustep_base_NSObject_hashCode (JNIEnv *env, jobject this)
{   
  id we;
  jint return_value;
  
  JIGS_ENTER;

  we = JIGSIdFromThis (env, this);
  return_value = (jint)[we hash];
  
  JIGS_EXIT_WITH_FAIL_VALUE (-1);

  return return_value;
}


JNIEXPORT jobject JNICALL 
Java_gnu_gnustep_base_NSObject_clone (JNIEnv *env, jobject this)
{
  id thisObject;
  id newObject;
  jobject newProxyObject;
  
  JIGS_ENTER;

  thisObject = JIGSIdFromThis (env, this);
  newObject = [thisObject copy];
  newProxyObject = JIGSCreateNewJavaProxy (env, newObject);
  
  JIGS_EXIT_WITH_FAIL_VALUE (NULL);

  return newProxyObject;
}

JNIEXPORT jobject JNICALL 
Java_gnu_gnustep_base_NSObject_mutableClone (JNIEnv *env, jobject this)
{
  id thisObject;
  id newObject;
  jobject newProxyObject;
  
  JIGS_ENTER;

  thisObject = JIGSIdFromThis (env, this);
  newObject = [thisObject mutableCopy];
  newProxyObject = JIGSCreateNewJavaProxy (env, newObject);
  
  JIGS_EXIT_WITH_FAIL_VALUE (NULL);

  return newProxyObject;
}

JNIEXPORT jstring JNICALL 
Java_gnu_gnustep_base_NSObject_toString (JNIEnv *env, jobject this)
{
  id thisObject;
  NSString *description;
  jstring result;
  
  JIGS_ENTER;
  
  thisObject = JIGSIdFromThis (env, this);
  description = [thisObject description];
  result = JIGSJobjectFromId (env, description);
  
  JIGS_EXIT_WITH_FAIL_VALUE (NULL);
  
  return result;
}

JNIEXPORT void JNICALL 
Java_gnu_gnustep_base_NSObject_retainObject 
(JNIEnv *env, jclass this, jobject object)
{
  id objc;
  
  JIGS_ENTER;
 
  objc = JIGSIdFromJobject (env, object);
  RETAIN (objc);

  JIGS_EXIT;  
}


JNIEXPORT void JNICALL 
Java_gnu_gnustep_base_NSObject_releaseObject 
(JNIEnv *env, jclass this, jobject object)
{
  id objc;
  
  JIGS_ENTER;
  
  objc = JIGSIdFromJobject (env, object);
  RELEASE (objc);
    
  JIGS_EXIT;  
}

File: Java/gnu/gnustep/base/NSObject.java
/* NSObject.java - the base proxy class
   Copyright (C) 2000 Free Software Foundation, Inc.

   Written by:  Nicola Pero <nicola@brainstorm.co.uk>
   Date: June 2000
   
   This file is part of the GNUstep Java Interface Library.

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU Library General Public
   License as published by the Free Software Foundation; either
   version 2 of the License, or (at your option) any later version.
   
   This library is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Library General Public License for more details.
   
   You should have received a copy of the GNU Library General Public
   License along with this library; if not, write to the Free
   Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111 USA.  */
package gnu.gnustep.base;

import gnu.gnustep.java.JIGS;

public class NSObject implements Cloneable
{
  /*
   * This integer is, in practice, a C pointer to the real object that 
   * this objects proxies.
   * We need to access it directly from native code; but native code 
   * does not care about the private modifier.
   * It should never be accessed from java (except from constructors), 
   * and that is why it is private.
   */
  private long realObject;
  
  static 
  {
    JIGS.initialize ();
  }
  
  /*
   * We can't implement a constructor as native. 
   * So, each constructor should be paired to a method constructor_native, 
   * with the same args, and implemented as native. 
   * The implementation of the constructor should invoke super (), 
   * then set this.realObject to the result of the constructor_native 
   * call, exactly as in the following example.
   */
  public NSObject ()
  {
    super ();
    this.realObject = this.NSObject_native ();
  }

  private native long NSObject_native ();

  /*
   * A similar trick with finalize; this is specific to NSObject.
   */
  protected void finalize () 
    throws Throwable
  {
    this.finalize_native (this.realObject);
    super.finalize ();
  }

  private native void finalize_native (long pointer)
    throws Throwable;
  
  /*
   * Here we have the exposed NSObject methods 
   */

  public native boolean equals (Object anObject);

  public native int hashCode ();

  public native Object clone()
    throws java.lang.CloneNotSupportedException;

  public native Object mutableClone ();

  public native String toString ();

  public String description ()
  {
    return this.toString ();
  }

  // (apple) native public Object valueForKey (String keyName);
  // (apple) native public void takeValueForKey (Object anObject, 
  //                                             String keyName);   

  /*
    GNUstep specific extension 
  */

  /* Use the following methods to deal transparently with weak
     references passed to Objective-C methods.  For example, if you want to
     set your java object 'javaDelegate' as delegate of the objective-C
     object 'objcObject', do as follows:

       NSObject.retainObject (javaDelegate);
       objcObject.setDelegate (javaDelegate);

     If you don't do this (or don't retain in some other way javaDelegate
     on the objective-C side before passing it to setDelegate), your
     program will segfault.  This is not a bug in the JIGS: it can't be
     otherwise.

     Warning: you need to 'retain' javaDelegate _before_ passing it through
     the JIGS.  But you don't need to retain javaDelegate if it already
     passed through the JIGS and some objective-C object is already keeping
     a reference to it.

     After you are done with the object (for example, when you set a
     different delegate), you should `release' it, as in the following
     example which changes the delegate of objcObject:
      
       NSObject.retainObject (newJavaDelegate);
       objcObject.setDelegate (newJavaDelegate);
       NSObject.releaseObject (javaDelegate);

     The rule is: you 'retain' the object before setting it as delegate,
     and 'release' it after having set another delegate.  
   
     Normally, you don't need these methods, they are needed only when
     calling Objective-C methods which store a reference to the argument
     without retaining it, such 'setDelegate:' and 'setTarget:'. */

  static native public void retainObject (Object object);
  static native public void releaseObject (Object object);
}
